### DISCUSSION TITLE
open-webui built-in API quick usage guide - OpenAI compatible ollama endpoint vs. open-webui endpoint

### DISCUSSION THREAD

**Discussion by jojasadventure** [open-webui/open-webui#5033] [General]
After a while of puzzling this together, I asked Claude to summarize. Please correct any mistaken information. Also, in case it helps anyone:

# OpenWebUI and Ollama API Usage Guide

This guide explains how to interact with the OpenWebUI and Ollama APIs, including the challenges and limitations currently faced by developers.

## Overview

OpenWebUI serves two main functions:
1. A frontend interface for Ollama
2. A proxy and security filter for API access to Ollama

There are two sets of API endpoints to be aware of:

1. OpenWebUI Management API
2. Ollama API (proxied through OpenWebUI)

### OpenWebUI Management API

Base URL: `http://your-openwebui-url/api/`

These endpoints are for managing OpenWebUI itself (see standard /api/v1/docs endpoint):

- `configs`
- `auths`
- `users`
- `chats`
- `documents`
- `models`
- `prompts`
- `memories`
- `files`
- `tools`
- `functions`
- `utils`

Note: These endpoints are primarily for managing the WebUI features and don't directly interact with the language model.

### Ollama API (proxied through OpenWebUI)

Base URL: `http://your-openwebui-url/ollama/`

Some endpoints (see ollama API docs for more)
- `/api/chat`: For chat completions
- `/api/tags`: For fetching available models

These endpoints interact directly with Ollama for model operations and chat completions.

### Installation

Ensure you have the `requests` library installed:

```
pip install requests
```

### Initializing the API client

```python
from ollama_api import OllamaAPI

api = OllamaAPI(base_url="http://your-openwebui-url", api_key="your-api-key")
```

### Getting available models

```python
models = api.get_models()
print(models)
```

This will return a list of dictionaries, each containing the `id` and `object` of a model.

### Sending a chat completion request

```python
messages = [
    {"role": "user", "content": "Hello, how are you?"}
]
response = api.chat_completion(model="model-name", messages=messages)
print(response)
```

This will return a dictionary with the following structure:

```python
{
    "id": "chatcmpl-X",
    "object": "chat.completion",
    "created": 1630000000,
    "model": "model-name",
    "choices": [
        {
            "index": 0,
            "message": {
                "role": "assistant",
                "content": "Response content here"
            },
            "finish_reason": "stop"
        }
    ],
    "usage": {
        "prompt_tokens": -1,
        "completion_tokens": -1,
        "total_tokens": -1
    }
}
```

## Advanced Features (RAG and Tools)

OpenWebUI offers Retrieval-Augmented Generation (RAG) and tool use capabilities, but using these features programmatically through the API is challenging due to limited documentation.

Current limitations:
- The OpenWebUI Management API endpoints (like `documents`, `tools`, `functions`) are visible in the Swagger documentation, but direct completion endpoints for RAG or tool use are not apparent in the API documentation.
 

## Challenges and Limitations

- The documentation for the OpenWebUI API is currently lacking, especially for advanced features.
- [Developers are resorting to examining code and commit history to understand API usage.](https://hostkey.com/blog/72-lets-build-a-customer-support-chatbot-using-rag-and-your-companys-documentation-in-openwebui/)
- While basic model calls are possible, implementing RAG functionality through the API has proven difficult. 

## Error Handling

The `get_models()` and `chat_completion()` methods will raise exceptions if there are any HTTP errors or if the API returns an error response. Make sure to wrap your API calls in try-except blocks to handle these exceptions gracefully.

**Comment by mdlmarkham** [comment]
Thank you! I'm trying to use OpenWebUI as a proxy for several instances of Ollama... I've tried almost every combination of openwebui:3000/ollama/ that I can think of. Keep getting 404 or 405 errors.

**Reply by jojasadventure** [reply]
I should have included that I am using the docker version which includes ollama directly, apologies. So I didn't run into any of these.


> Thank you! I'm trying to use OpenWebUI as a proxy for several instances of Ollama... I've tried almost every combination of openwebui:3000/ollama/ that I can think of. Keep getting 404 or 405 errors.

**Comment by c9482** [comment]
This is great stuff, thanks! I wanted to add that navigating `http://your-openwebui-url/ollama/docs` will display FastAPI page with all methods supported by the API.

My question though is whether making a reference to #embeddeddocument in system prompt when making a chat completion will use it as knowledge base. Or is the chat completion passed directly to Ollama and Ollama is not aware of the embedded document?

**Reply by jojasadventure** [reply]
Neat, I had totally omitted the fastAPI docs page. Thanks. 

I haven't looked through the code to see how that works. Probably much quicker just to try it, to find out... I wouldn't hold much hope. The documentation seems pretty clear about it being a proxy and a security filter...

**Comment by kklasmeier** [comment]
I customized my own model through the workspace in Open WebUI called "offer-configuration-assistant". I want to use that to chat via an application, so i want to Programmatically make calls through an API. 

Starting with curl, i thought i would try variations of the below, but without success. Does anyone know the endpoint and definition of chats to models created in workspaces?

curl -X POST -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjIwNzU2MGIyLTU1MGMtNDgwMS1hZjYyLTFhN2I2MmM5ZThkNyJ9.E6e6zfFk9tD5FACOSkeARwDSmBCoph32tMJWlPifOPw" -H 'Content-Type: application/json' http://192.168.1.7:3000/ollama/api/generate -d '{ "model": "offer-configuration-assistant:latest", "prompt": "Why is the sky blue?", "stream": false }'

{"detail":"Model 'offer-configuration-assistant:latest' was not found"}

**Reply by jojasadventure** [reply]
afaik it's not happening (yet) sorry

**Reply by c9482** [reply]
This worked for me:

```
curl 'https://<my openwebui url>/ollama/api/chat' \
  -H 'accept: */*' \
  -H 'authorization: Bearer sk-xxxxxxxxxxxxx' \
  -H 'content-type: application/json' \
  --data-raw $'{"stream":false,"model":"my-custom-model","messages":[{"role":"user","content":"who are you?"}]}'
```
  
Response:

`{"model":"llama3.1:latest","created_at":"2024-09-05T16:15:46.531724Z","message":{"role":"assistant","content":"Hello! I'm here to help with any questions or concerns you may have regarding the XXXXXXXXX. What would you like to know? Have you reviewed the ordinance and need clarification on a specific section? Or perhaps you're looking for guidance on how to apply its regulations in a particular situation? I'm here to assist!"},"done_reason":"stop","done":true,"total_duration":1552237334,"load_duration":35436250,"prompt_eval_count":70,"prompt_eval_duration":211511000,"eval_count":79,"eval_duration":1303703000}`

Notice that the response contained "llama3.1:latest" which is the base model for my custom model. Also, the response was relevant to system prompt set for my custom model in OpenwebUI.

However, running "ollama list" does not display my custom model. So it looks like OpenwebUI is adding information from the custom model when it's passing it to Ollama.

**Comment by eloyestebann51** [comment]
Hello, as my colleague mentioned before, I want to use my own model for a website. I have a plugin to which I want to implement AI with an API. I'm a bit lost, the idea is that the user on the web can talk to the AI, just text messages. If you can give me some guidance, thank you.

**Comment by mbond99** [comment]
Hello, I wanted to post this here in case it might help someone out! 

I have been experimenting with AI agents using the autogen python library, and wanted to connect to the custom models I created in Open Web UI. While you can customize a model and RAG through the autogen library directly, my coworker and I wanted to keep as much of our model customization and RAG knowledge bases as possible in a central location (Open Web UI). The key to getting this to work was using the base URL "http://OpenWebUIURL:port/api" (ex. http://localhost:3000/api). I imagine this URL would work with anything that allows for an OpenAI API connection for other use cases!

My config for autogen is below if that's of interest to anyone:
OpenAIChatCompletionClient(
        model="test-model",
        api_key="sk-xxxxxxxxxxxx",
        base_url="http://localhost:3000/api",
        model_capabilities={
            "json_output": False,
            "vision": False,
            "function_calling": True,
        },
        timeout=120,
    )

I am having issues with tool calling using this method. I appreciate any suggestions if anyone has an idea about how to fix this!

**Comment by phazethru** [comment]
I was trying to connect HomeAssistant to an Open WebUI model connected to llamacpp and came across this page.  In case you're trying to do the same, you'll want to:

1. Enable OpenAI API in Open WebUI, and configure it to point to your llamacpp
2. Create a model and configure as you like
3. Configure an API Key under authentication
4. Use the llamacpp homeassistant integration with the API Key, and this Open WebUI endpoint: https://[your url>]/api/

If you try and use  https://[your url]/openai/  it will only do passthru and not show your models.
